sling/nlp/wiki/wikipedia-links.cc

```diff
@@ -0,0 +1,205 @@
+// limitations under the License.
+
+#include "sling/base/logging.h"
+#include "sling/base/types.h"
+#include "sling/nlp/document/document.h"
+#include "sling/nlp/document/document-tokenizer.h"
+#include "sling/nlp/document/lex.h"
+#include "sling/task/documents.h"
+#include "sling/task/accumulator.h"
+#include "sling/task/reducer.h"
+
+namespace sling {
+namespace nlp {
+
+class WikipediaLinkExtractor : public task::DocumentProcessor {
+ public:
+  void Startup(task::Task *task) override {
+    // Get parameters.
+    task->Fetch("extract_mention_links", &extract_mention_links_);
+    task->Fetch("extract_theme_links", &extract_theme_links_);
+    task->Fetch("extract_infobox_links", &extract_infobox_links_);
+
+    // Initialize fan-in output.
+    fanin_ = task->GetSink("fanin");
+    if (fanin_ != nullptr) {
+      fanin_counts_.Init(fanin_);
+    }
+
+    // Statistics.
+    num_links_ = task->GetCounter("links");
+    num_mention_links_ = task->GetCounter("mention_links");
+    num_theme_links_ = task->GetCounter("theme_links");
+  }
+
+  void Process(Slice key, const Document &document) override {
+    // Collect outbound links from document.
+    HandleMap<int> links;
+    Store *store = document.store();
+    ExtractDocumentLinks(document, &links);
+
+    if (!links.empty()) {
+      // Output frame with link statistics.
+      Builder b(store);
+      for (auto it : links) {
+        b.Add(it.first, it.second);
+      }
+      FrameProcessor::Output(key, b.Create());
+
+      // Accumulate fan-in statistics.
+      if (fanin_ != nullptr) {
+        for (auto it : links) {
+          Frame target(store, it.first);
+          fanin_counts_.Increment(target.Id(), it.second);
+        }
+      }
+    }
+  }
+
+  void Flush(task::Task *task) override {
+    if (fanin_ != nullptr) {
+      fanin_counts_.Flush();
+    }
+  }
+
+  void ExtractDocumentLinks(const Document &document, HandleMap<int> *links) {
+    // Collect all links in mentions.
+    Store *store = document.store();
+    if (extract_mention_links_) {
+      Handles evoked(store);
+      for (const Span *span : document.spans()) {
+        span->AllEvoked(&evoked);
+        for (Handle link : evoked) {
+          link = store->Resolve(link);
+          if (!store->IsFrame(link)) continue;
+          if (store->GetFrame(link)->IsNamed()) {
+            (*links)[link]++;
+            num_mention_links_->Increment();
+            num_links_->Increment();
+          }
+        }
+      }
+    }
+
+    // Collect all thematic links.
+    for (Handle link : document.themes()) {
+      link = store->Resolve(link);
+      if (store->IsFrame(link)) {
+        if (store->GetFrame(link)->IsNamed()) {
+          if (extract_mention_links_) {
+            (*links)[link]++;
+            num_theme_links_->Increment();
+            num_links_->Increment();
+          }
+        } else {
+          if (extract_infobox_links_) {
```

Comment:
Is it guaranteed that all unnamed thematic frames are coming from infoboxes?
If there is an isa slot in 'f' that denotes it is coming from an infobox, maybe we should check for that slot here?

Reply:
That's a good point. I have added a check for `isa:/wp/infobox`.